WIP: Reduce allocations under RemoveAssetProvider.SynchronizeAssetsAsync

ToddGrun opened the pull request:
This code realized an ImmutableArray before from an ArrayBuilder. This immutablearray was only used within the scope of SynchronizeAssetsAsync.

*** Prior allocations from speedometer:
![image](https://github.com/dotnet/roslyn/assets/6785178/bdec15d1-ab2c-44af-a483-3c3ab90e1a46)

CyrusNajmabadi commented on src/Workspaces/Remote/ServiceHub/Host/AssetProvider.cs:
```diff
@@ -26,6 +27,8 @@ internal sealed partial class AssetProvider(Checksum solutionChecksum, SolutionA
     private readonly SolutionAssetCache _assetCache = assetCache;
     private readonly IAssetSource _assetSource = assetSource;
 
+    private static readonly ObjectPool<SegmentedList<Checksum>> s_checksumListPool = new ObjectPool<SegmentedList<Checksum>>(() => new SegmentedList<Checksum>());
```
```suggestion
    private static readonly ObjectPool<SegmentedList<Checksum>> s_checksumListPool = new(() => new());
```

CyrusNajmabadi commented on src/Workspaces/Remote/ServiceHub/Host/AssetProvider.cs:
```diff
@@ -118,28 +121,32 @@ public async ValueTask SynchronizeProjectAssetsAsync(ProjectStateChecksums proje
         using (Logger.LogBlock(FunctionId.AssetService_SynchronizeAssetsAsync, Checksum.GetChecksumsLogInfo, checksums, cancellationToken))
         {
-            using var _1 = ArrayBuilder<Checksum>.GetInstance(checksums.Count, out var missingChecksums);
+            var missingChecksums = s_checksumListPool.Allocate();
```
can you use .GetPooledObject() instead?  then you can use this in a `using` and be sure to return this even if canceled.

ToddGrun replied:
I believe that would called the pooledobjject's releaser, which will do the size threshold check.

ToddGrun replied:
Ooh, there is a trimOnFree param, let me try that out

ToddGrun replied:
Works wonderfully, thanks!

sharwell commented on src/Workspaces/Core/Portable/Serialization/PooledList.cs:
```diff
@@ -15,7 +14,7 @@ namespace Microsoft.CodeAnalysis.Serialization
     internal static class Creator
     {
-        public static PooledObject<HashSet<Checksum>> CreateChecksumSet(ImmutableArray<Checksum> checksums)
+        public static PooledObject<HashSet<Checksum>> CreateChecksumSet(SegmentedList<Checksum> checksums)
```
We don't want a mutable type here.

```suggestion
        public static PooledObject<HashSet<Checksum>> CreateChecksumSet(ImmutableSegmentedList<Checksum> checksums)
```

ToddGrun replied:
I'd love for this to be an immutable type, but I don't see how to get one down to this level without allocating up in SynchronizeAssetsAsync. Maybe I'm missing a trick?

CyrusNajmabadi replied:
@ToddGrun what about, as an alternative, a readonly struct wrapper around SegmentedList?

ToddGrun replied:
I'm not clear on what that would gain. Yes, the wrapper would be immutable, but I assume it would give access to the mutable structure. I don't mind the current approach, especially since it shows promising allocation results, but am willing to consider other approaches.

sharwell replied:
> a readonly struct wrapper around SegmentedList?

`ImmutableSegmentedList<T>` is exactly that.

sharwell commented on src/Workspaces/Remote/ServiceHub/Host/AssetProvider.cs:
```diff
@@ -118,28 +122,29 @@ public async ValueTask SynchronizeProjectAssetsAsync(ProjectStateChecksums proje
     {
-        Contract.ThrowIfTrue(checksums.Contains(Checksum.Null));
-        if (checksums.Length == 0)
+        Contract.ThrowIfTrue((checksums as ICollection<Checksum>).Contains(Checksum.Null));
```
`Contract.ThrowIfTrue` is not a debug-only operation. Are we sure (for `ImmutableSegmentedList<T>`) that this isn't introducing a boxing operation?

ToddGrun replied:
I don't remember why I needed to do this originally, but it looks like SegmentedList implements Contains directly, so this cast shouldn't be necessary any more.

sharwell commented on src/Workspaces/Remote/ServiceHub/Host/AssetProvider.cs:
```diff
@@ -118,28 +122,29 @@ public async ValueTask SynchronizeProjectAssetsAsync(ProjectStateChecksums proje
         using (Logger.LogBlock(FunctionId.AssetService_SynchronizeAssetsAsync, Checksum.GetChecksumsLogInfo, checksums, cancellationToken))
         {
-            using var _1 = ArrayBuilder<Checksum>.GetInstance(checksums.Count, out var missingChecksums);
+            using var missingChecksumsObject = s_checksumListPool.GetPooledObject();
```
This intermediate collection is completely unnecessary. Just use the `checksums` argument directly:

```cs
// Remove all values from 'checksums' where the asset cache already contains a value
checksums.RemoveWhere(checksum => _assetCache.TryGetAsset<object>(checksum, out _));
```

This can be further optimized by defining `PooledDelegates.GetPooledPredicate`, if necessary:

```cs
// Remove all values from 'checksums' where the asset cache already contains a value
using (var _1 = PooledDelegates.GetPooledPredicate<Checksum, AssetProvider>(
    static (checksum, arg) => arg._assetCache.TryGetAsset<object>(checksum, out _),
    this,
    out var predicate))
{
    checksums.RemoveWhere(predicate);
}
```

ToddGrun replied:
I don't think I'm following this suggestion. 

Modifying that hashset in place requires knowledge that the caller is ok with that. It might be ok, but I'd need to verify that. 

Assuming that were ok, then I'd need to pass that HashSet down and change all the calls to take in a HashSet instead of a SegmentedList, right?

sharwell replied:
> Modifying that hashset in place requires knowledge that the caller is ok with that. It might be ok, but I'd need to verify that.

I did verify it, but we'd want a way to make it apparent to the callers.

> Assuming that were ok, then I'd need to pass that HashSet down and change all the calls to take in a HashSet instead of a SegmentedList, right?

That's one option. We could also create an `ImmutableArray` from it directly by calling `ImmutableArray.CreateRange<T>`, without going through the builder pool.

ToddGrun replied:
The intent of this PR was to investigate removing the allocation of creating that ImmutableArray. If we *need* that data structure to be passed down as immutable, I should just close out this PR. But, from what I saw, it didn't need to be immutable, it was just a nice to have. This is an investigation into whether removing that nice to have could have a positive perf impact without too much ugliness in the code.

As for the suggestion to modify the HashSet in-place, I guess I'm still not seeing the value. Perhaps the speedometer profile with this change will show that's an issue, but I'd prefer to not make that change unless it's needed as it does change the semantics around the hashset passed into this method.

CyrusNajmabadi replied:
as these are all internal components, without a surface area exposed to other features, i would be fine with this not being immutable, esp if it cuts down on churn.

ToddGrun commented:
This looks effective from the [speedometer run](https://devdiv.visualstudio.com/DevDiv/_apps/hub/ms-vseng.pit-vsengPerf.pit-hub?targetBuild=34622.165.dn-bot.240223.030609.530878&targetBranch=34622.165.dn-bot.240223.030609.530878&targetPerfBuildId=9123365&runGroup=Speedometer&since=2024-02-23&baselineBuild=34622.165&baselineBranch=main) here (CompletionTest.Completion/9990.Totals/CLR_BytesAllocates_NonDevenv)

```
Name                                   	Exc %	       Exc
Type Microsoft.CodeAnalysis.Checksum[] 	  0.1	 4,900,776
```

No HashSet\<Checksum\> appears in the profile.

Going to elevate out of draft status

ToddGrun commented:
@sharwell, @CyrusNajmabadi -- I've moved this out of draft status as it shows promise, ptal

ToddGrun commented:
> Requesting changes pending investigation on not losing argument immutability

Would like to ping @CyrusNajmabadi and @sharwell for preferences on how to proceed. I personally don't mind sending the mutable types through the callstack as all the bits they proceed through are internal. 

As I know there is concern with that approach, I investigated what would be needed to make them immutable (ImmutableSegmentedList instead of ImmutableArray). It can be done ([branch here](https://github.com/ToddGrun/roslyn/compare/dev/toddgrun/SynchronizeAssetsAsync_Allocations...ToddGrun:roslyn:dev/toddgrun/SynchronizeAssetsAsync_Allocations_ImmutableCollectionsMarshal)), but it's a bit more effort:

1) Added msgpack serialization for ImmutableSegmentedList
2) Added ImmutableCollectionsMarshal equivalent for ImmutableSegmentedList

I'm not in love with these changes, but I couldn't come up with an approach for sending an immutabletype through without allocations otherwise.

ToddGrun commented:
> Requesting changes pending investigation on not losing argument immutability

@sharwell -- I'd like to keep this moving along. Do you think the ValueBuilder PR will go in sometime soon, or should I plan on making progress on this PR without that? I'm mostly interested in the ImmutableCollectionsMarshal equivalent that this PR created a rudimentary version of. (I'm also intereseted in the ImmutableSegmentedDictionary.AddRange optimization that I merged 72220 without)

sharwell commented on src/Dependencies/Collections/ImmutableSegmentedList`1.cs:
```diff
@@ -75,6 +75,14 @@ namespace Microsoft.CodeAnalysis.Collections
 
         private readonly SegmentedList<T> _list;
 
+        public static class ImmutableCollectionsMarshal
+        {
+            public static ImmutableSegmentedList<T> ToImmutableSegmentedList(SegmentedList<T> list)
+            {
+                return new ImmutableSegmentedList<T>(list);
+            }
+        }
```
This can be removed now

sharwell commented on src/Workspaces/Remote/Core/Serialization/ImmutableSegmentedListFormatter.cs:
```diff
@@ -0,0 +1,60 @@
+
+        var formatter = options.Resolver.GetFormatterWithVerify<T>();
+
+        // TODO: Should there be a CreateBuilder that takes in an initial capacity
+        // similar to what ImmutableArray allows?
```
Yes, there should. In the meantime, you can use `SegmentedCollectionsMarshal.AsImmutableSegmentedList` if you want.

sharwell replied:
I have a branch ready to submit for this as soon as #62193 merges.
